geode-core/src/test/java/org/apache/geode/cache/query/internal/AttributeDescriptorTest.java

```diff
@@ -232,18 +307,91 @@ public void readReflectionForImplicitMethodInvocationShouldThrowExceptionWhenMet
 
-    assertThatThrownBy(() -> attributeDescriptor.readReflection(testBean))
+    assertThatThrownBy(() -> attributeDescriptor.readReflection(testBean, queryExecutionContext))
+        .isInstanceOf(NotAuthorizedException.class)
+        .hasMessageStartingWith(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING);
+  }
+
+  @Test
+  @Parameters({
+      "nonPublicAttributeWithPublicAccessor, nonPublicAttributeWithPublicAccessor",
+      "nonPublicAttributeWithPublicGetterMethod, getNonPublicAttributeWithPublicGetterMethod"})
+  public void readReflectionForImplicitMethodInvocationShouldShouldThrowExceptionWhenMethodIsNotAuthorizedAndCacheResult(
+      String attributeName, String methodName) {
+    doReturn(false).when(methodInvocationAuthorizer).authorize(any(), any());
+    AttributeDescriptor attributeDescriptor =
+        new AttributeDescriptor(typeRegistry, methodInvocationAuthorizer, attributeName);
+
+    assertThatThrownBy(() -> attributeDescriptor.readReflection(testBean, queryExecutionContext))
         .isInstanceOf(NotAuthorizedException.class)
         .hasMessageStartingWith(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING);
+
+    String cacheKey = TestBean.class.getCanonicalName() + "." + methodName;
+    assertThat(queryExecutionContext.cacheGet(cacheKey)).isNotNull();
+    assertThat(queryExecutionContext.cacheGet(cacheKey)).isEqualTo(false);
+  }
+
+  @Test
+  public void readReflectionForImplicitMethodInvocationShouldUseAlreadyAuthorizedCachedResultWhenMethodIsForbiddenAndQueryContextIsTheSame() {
+    doReturn(false).when(methodInvocationAuthorizer).authorize(any(), any());
+    AttributeDescriptor nonPublicAttributeWithPublicAccessor =
+        new AttributeDescriptor(typeRegistry, methodInvocationAuthorizer,
+            "nonPublicAttributeWithPublicAccessor");
+    AttributeDescriptor nonPublicAttributeWithPublicGetterMethod =
+        new AttributeDescriptor(typeRegistry, methodInvocationAuthorizer,
+            "nonPublicAttributeWithPublicGetterMethod");
+
+    // Same QueryExecutionContext -> Use cache.
+    IntStream.range(0, 10).forEach(element -> {
+      assertThatThrownBy(() -> nonPublicAttributeWithPublicAccessor
+          .readReflection(testBean, queryExecutionContext))
+              .isInstanceOf(NotAuthorizedException.class)
+              .hasMessageStartingWith(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING);
+
+      assertThatThrownBy(() -> nonPublicAttributeWithPublicGetterMethod
+          .readReflection(testBean, queryExecutionContext))
+              .isInstanceOf(NotAuthorizedException.class)
+              .hasMessageStartingWith(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING);
+    });
+    verify(methodInvocationAuthorizer, times(2)).authorize(any(), any());
+  }
+
+  @Test
+  public void readReflectionForImplicitMethodInvocationShouldNotUseAlreadyAuthorizedCachedResultWhenMethodIsForbiddenAndQueryContextIsNotTheSame() {
```

Comment:
See above comment regarding "already authorized" in test name.

Reply:
👍